cpp/Platform.Delegates/Platform.Delegates.Delegate.h

```diff
@@ -16,54 +14,134 @@ namespace Platform::Delegates
 
+        constexpr Delegate() noexcept = default;
+
+        Delegate(const Delegate&) noexcept = default;
+
+        Delegate(Delegate&&) noexcept = default;
+
+        constexpr Delegate(DelegateRawFunctionType simpleFunction) noexcept
+                : simpleFunction(simpleFunction) {}
+
+        Delegate(const DelegateFunctionType &complexFunction)
+                : complexFunction(std::make_shared<DelegateFunctionType>(complexFunction)) {}
+
+        template <typename Class>
+        Delegate(std::shared_ptr<Class> object, ReturnType(Class:: *member)(Args...))
+                : Delegate(std::make_shared<MemberMethod<Class>>(std::move(object), member)) { }
+
+        virtual ~Delegate() = default;
+
+        Delegate &operator=(const Delegate &other) noexcept
+        {
+            if (this != &other)
+            {
+                this->simpleFunction = other.simpleFunction;
+                this->memberMethod = other.memberMethod;
+                this->complexFunction = other.complexFunction;
+            }
+            return *this;
+        }
+
+        Delegate &operator=(Delegate&& other) noexcept {
+            if (this != &other)
+            {
+                this->simpleFunction = std::move(other.simpleFunction);
+                this->memberMethod = std::move(other.memberMethod);
+                this->complexFunction = std::move(other.complexFunction);
+            }
+            return *this;
+        }
+
+        virtual ReturnType operator()(Args... args)
+        {
+            if (simpleFunction)
+            {
+                return simpleFunction(args...);
+            }
+            if (memberMethod)
+            {
+                return (*memberMethod)(args...);
+            }
+            if (complexFunction)
+            {
+                return (*complexFunction)(args...);
+            }
+
+            throw std::bad_function_call{};
+        }
+
+        virtual bool operator==(const Delegate &other) const
+        {
+            if (simpleFunction && other.simpleFunction)
+            {
+                return simpleFunction == other.simpleFunction;
+            }
+            if (memberMethod && other.memberMethod)
+            {
+                return *memberMethod == *other.memberMethod;
+            }
+            if (complexFunction && other.complexFunction)
+            {
+                return AreFunctionsEqual(*complexFunction, *other.complexFunction);
+            }
+            return false;
+        }
+
+        virtual bool operator!=(const Delegate &other) const
+        {
+            return !(*this == other);
+        }
+    private:
         class MemberMethodBase
         {
         public:
             virtual ReturnType operator()(Args... args) = 0;
             virtual bool operator== (const MemberMethodBase &other) const = 0;
             virtual bool operator!= (const MemberMethodBase &other) const = 0;
+            virtual ~MemberMethodBase() = default;
```

Comment:
Why we need default constructor here?

Reply:
It is a destructor ;) We declare it explicitly to indicate that it is virtual. This will help avoid memory leaks in this case:
```c++
struct A {
};

struct B : public A {
    int * ptr;
    /* code */
    ~B() { delete ptr; }
};

void foo() {
    A *a = new B();
    // code
    delete a; // ~ B will not be called because the destructor ~ A is called
}
```
You can see more detailed explanations here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c35-a-base-class-destructor-should-be-either-public-and-virtual-or-protected-and-non-virtual

Reply:
Ah, ok.